schema.c

```diff
@@ -4065,3 +4367,553 @@ sch_json_to_gnode (sch_instance * instance, sch_node * schema, json_t * json, in
+        if (ret)
+            sch_lua_error (instance->ls, ret);
+        else if (lua_gettop (instance->ls) != (s_0 + 2))
+        {
+            syslog (LOG_ERR, "XML LUA: Stack not zero(%d) after function: %s.%s\n",
+                    lua_gettop (instance->ls), xlat_data->lua_module,
+                    xlat_data->lua_func);
+            lua_pop (instance->ls, 1);
+        }
+        else
+        {
+            const char *value = lua_tostring (instance->ls, -1);
+            char *old_value = (char *) node->children->data;
+            node->children->data = g_strdup (value);
+            DEBUG (flags, "function: %s old value %s new value %s\n", xlat_data->lua_func, old_value, value);
+            g_free (old_value);
+        }
+    }
+    else
+    {
+        lua_pop(instance->ls, 1);
+        return;
+    }
+}
+
+static bool
+sch_translate_child_exists (GNode *parent, char *name, GNode **ret_node)
+{
+    GNode *child;
+
+    for (child = parent->children; child; child = child->next)
+    {
+        if (g_strcmp0 (APTERYX_NAME (child), name) == 0)
+        {
+            *ret_node = child;
+            return true;
+        }
+    }
+
+    *ret_node = NULL;
+    return false;
+}
+
+static bool
+sch_translate_sibling_exists (GNode *node, char *name, GNode **ret_node)
+{
+    while (node)
+    {
+        if (g_strcmp0 (APTERYX_NAME (node), name) == 0)
+        {
+            *ret_node = node;
+            return true;
+        }
+        node = node->next;
+    }
+
+    *ret_node = NULL;
+    return false;
+}
+
+static bool
+xlat_tree_match (GNode *tree, GNode *pattern, GNode **last_match_node)
+{
+    bool match = false;
+    GNode *node1 = tree;
+    GNode *node2 = pattern;
+    GNode *ret_node;
+    GNode *last_valid_node = NULL;
+
+    while (node1 && node2)
+    {
+        match = true;
+        if (!sch_translate_sibling_exists (node1, APTERYX_NAME (node2), &ret_node) &&
+            g_strcmp0 ((char *) node2->data, "*") != 0)
+            return false;
+        if (ret_node)
+        {
+            last_valid_node = ret_node;
+            node1 = ret_node->children;
+            ret_node = NULL;
+        }
+        else
+        {
+            last_valid_node = node1;
+            node1 = node1->children;
+        }
+        node2 = node2->children;
+    }
+
+    if (node2)
+        match = false;
+    else
+        *last_match_node = last_valid_node;
+
+    return match;
+}
+
+static char *
+sch_translate_get_wildcard_data (GNode *node, GNode *tree, uint32_t pos)
+{
+    GNode *nnode = tree;
+    uint32_t matches = 0;
+
+    while (nnode && node)
+    {
+        if (g_strcmp0 (APTERYX_NAME (nnode), "*") == 0)
+        {
+            matches++;
+            if (matches == pos)
+                return g_strdup (APTERYX_NAME (node));
+        }
+
+        nnode = nnode->children;
+        node = node->children;
+    }
+    return NULL;
+}
+
+static inline gboolean
+_sch_translate_count_nodes (GNode *node, void *data)
```

Comment:
feels like the is a glib fn for this

Reply:
This is a special counter. It does not count leaf nodes that do not have a valid data pointer. Such nodes appear on queries. The glib function "g_node_n_nodes" is used in the routine sch_translate_output